### ISSUE TITLE
Metaprotocol playlists - features

### ISSUE THREAD

**Issue by zeeshanakram3** [Joystream/joystream#3463]
This issues lists protobuf/graphql schemas, and validations on Playlists actions based on requirements in #3417

## Protobuf Message Schemas
```proto
// Create a playlist
message CreatePlaylist {
    // Associated channel's ID of the playlist
    required uint64 channel_id = 1;

    // Title of the playlist
    required string title = 2;

    // Description of the playlist
    required string description = 3;

    // IDs of the videos to include in playlist (in given order)
    repeated uint64 video_ids = 4;
    
    // Image uri of the playlist's thumbnail
    optional string thumbnail_uri = 5;

    // Playlist status, whether it is public or private. If the 
    // field is omitted the default playlist status would be public???
    optional bool is_public = 6;
}

// Update a playlist
message UpdatePlaylist {
    // ID of the playlist to be updated
    required string playlist_id = 1;

    // New title of the playlist (if any)
    optional string new_title = 2;

    // New description of the playlist (if any)
    optional string new_description = 3;

    // New image uri of the playlist's thumbnail (if any)
    optional string new_thumbnail_uri = 4;

    // IDs of the videos to be added in playlist (in given order) (if any)
    repeated uint64 new_video_ids = 5;
    
    // IDs of the videos to be removed from playlist (if any)
    repeated uint64 remove_video_ids = 6;

    // Video IDs in given order to update videos' order in the playlist 
    repeated uint64 new_order = 7;
    
    // Playlist status, whether it is public or private. 
    optional bool is_public = 8;
}

// Delete a playlist
message DeletePlaylist {
    // ID of the playlist to be deleted
    required string playlist_id = 1;
}


message ChannelOwnerRemarked {   
    oneof channel_owner_remarked {
        // Other channel owner remarked messages
        // ...
        CreatePlaylist create_playlist = 5;
        UpdatePlaylist update_playlist = 6;
        DeletePlaylist delete_playlist = 7;
    }
}
```
## Graphql Schemas
Excluding the event entities
```graphql
type Playlist @entity {
  "METAPROTOCOL-{network}-{blockNumber}-{indexInBlock}"
  id: ID!

  "Playlist associated channel"
  channel: Channel!

  "Title of the playlist"
  title: String!

  "Description of the playlist"
  description: String!

  # Many-to-many relationship between playlists and videos
  "List of videos in the playlist"
  videos: [Video!]! 

  "Number of public uncensored videos in the playlist"
  publicUncensoredVideosCount: Int!

  "Total duration of public uncensored videos in the playlist"
  publicUncensoredVideosDuration: Int!

  "Whether the Playlist is supposed to be publicly displayed"
  isPublic: Boolean

  "Playlist thumbnail asset (If not provided, the thumbnail of the first video will be used)"
  thumbnailPhoto: StorageDataObject
}

# Updated video entity
type Video @entity {
   # other fields
   ...
   
  "List of playlists the video is part of"
  videoInPlaylists: [Playlist!] @derivedFrom(field: "videos")
}

# Updated channel entity
type Channel @entity {
  # other fields
  ...
  
  "List of channel's playlists"
  channelPlaylists: [Playlist!] @derivedFrom(field: "channel")
}
```

# Playlists Actions
#### Protobuf message
`CreatePlaylist`
#### Event
`ChannelOwnerRemarked`
#### Validations
- _Ensure videos exist_
- _Ensure channel owns the videos_

## Update Playlist
In the current protobuf message design, multiple update actions can be performed simultaneously, i.e. changing the title or description along with adding or removing videos from the playlist, or doing all prior mentioned actions along with changing the order of videos in the playlist.
#### Protobuf message
`UpdatePlaylist`
#### Event
`ChannelOwnerRemarked`
#### Validations
- _Ensure playlist exists_
- _Ensure channel owns the playlist_
- _Ensure channel owns the videos in `new_videos_ids` (if any)_
- _**Side Effect: Append `new_videos_ids` to Playlist's `videos` field, in given order**_ 
- _Ensure channel owns the videos in `remove_videos_ids` (if any)_
- _Ensure Video IDs in `remove_videos_ids` are part of the playlist (if any)_
- _**Side Effect: Remove `remove_videos_ids` from Playlist's `videos` field, without changing the order of remaining videos**_ 
- _Ensure Video IDs in `new_order` are part of the playlist_
- _**Side Effect: Set playlist order according to `new_order`**_ 

#### Protobuf message
`DeletePlaylist`
#### Event
`ChannelOwnerRemarked`
#### Validations
- _Ensure playlist exists_
- _Ensure channel owns the playlist_

## Questions

From #3417 
> if the thumbnail is not provided, we would like the thumbnail of the playlist to be automatically populated with the thumbnail of the first video in the playlist.

Since [thumbnailPhoto](https://github.com/Joystream/joystream/blob/41d156c03ef648ec152593c6203d7f0b770575fe/query-node/schemas/content.graphql#L112) is a nullable field, if the thumbnail of the first video does not exist should we check for the thumbnail of the second video and so on.

***

Currently, all the playlist update operations are handled by a single protobuf message `UpdatePlaylist`, is it fine or should there be a dedicated message for each update action e.g. 
- `AddVideos` for adding videos to playlist
```proto
message AddVideos {
  required string playlist_id = 1;
  repeated uint64 video_ids = 2; 
}
```
- `RemoveVideos` for removing videos from playlist
```proto
message RemoveVideos {
  required string playlist_id = 1;
  repeated uint64 video_ids = 2; 
}
```

**Comment by bedeho** [comment]
Overall looks very solid, I'll leave @kdembler to verify further details, here are my thoughts:

1. `CreatePlaylist` seems to not allow creating a playlist with more than one video to begin with?
2. `UpdatePlaylist` seems overly complex in terms specifying changes, can't you just accept a new list that overwrites the existing one wholesale? thats the simplest most straight forward format.
3. Remember to also make entities for successful events/messages, which include things like playlist IDs, which the user did not provide as input.
4. @kdembler What should the query node do for thumbnail if the playlist is empty, or the first video has no thumbnail? can't this be done client side?
5. I think its fine with one message for updating.
6. How are playlist ids selected? is there one global id space?
7. When describing each message, I would suggest you separate validation and side-effects entirely. All validations must pass (also sometimes referred to as conditions or pre-conditions), before any side-effect can take place, right now its sort of mixed, like under `UpdatePlaylist`.

**Comment by zeeshanakram3** [comment]
> CreatePlaylist seems to not allow creating a playlist with more than one video to begin with?

That was a typo, I have updated the schema.

> UpdatePlaylist seems overly complex in terms specifying changes, can't you just accept a new list that overwrites the existing one wholesale? that's the simplest most straight forward format.

Yes, accepting a new list that overrides the existing one definitely seems a simpler solution, will go with this.

> How are playlist ids selected? is there one global id space?

Yes there is one global id space based `network`, `blockNumber` & `indexInBlock`.  Since Metaprotocol entities, i.e. _playlists, comments, reactions_, etc. don't have runtime existence (only exist in QN), there needs a way to assign a unique id across all entities of the same type. So I am using the following format to assign IDs to the playlist `"METAPROTOCOL-{network}-{blockNumber}-{indexInBlock}"`
This method is just the extension of the current method of assigning IDs to event entities (an [example](https://github.com/Joystream/joystream/blob/41d156c03ef648ec152593c6203d7f0b770575fe/query-node/schemas/councilEvents.graphql#L6)) in QN. i.e. currently IDs to event entities are assigned as `"{network}-{blockNumber}-{indexInBlock}"`

> When describing each message, I would suggest you separate validation and side-effects entirely. 

Got it.

@bedeho

**Comment by kdembler** [comment]
1. +1 from me on just overriding the full list of videos with new order, instead of separate add/remove/reorder.
2. One thing I was wondering about is ID space. Would it be hard to follow the same approach as runtime and use incremental integer IDs? The reason I'm asking is that we use those IDs in the URL, for example `video/1234`, `channel/1234`, etc. `playlist/METAPROTOCOL-OLYMPIA-123456-4` may be a bit unsexy 😅 
3. `CreatePlaylist` accepts `thumbnail_uri`, I would expect this to be a storage object to upload.
4. @bedeho re your question on thumbnails: if the playlist has no thumbnail, thumbnail of first video should be used. If there is no first video or that video doesn't have a thumbnail, it should be empty then. `can't this be done client side?` I guess you refer to falling back to first video thumbnail if playlist doesn't have one? It could be done client-side but then for every playlist we would need to run a separate query to get its first video. It will be much more convenient if this is available directly on the playlist object

**Comment by zeeshanakram3** [comment]
Closing as playlists were being implemented as runtime entities by overloading video metadata in https://github.com/Joystream/joystream/pull/3703
